pkg/yqlib/decoder_toml.go

```diff
@@ -0,0 +1,223 @@
+package yqlib
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+
+	toml "github.com/pelletier/go-toml/v2/unstable"
+	yaml "gopkg.in/yaml.v3"
+)
+
+type tomlDecoder struct {
+	parser   toml.Parser
+	finished bool
+	d        DataTreeNavigator
+}
+
+func NewTomlDecoder() Decoder {
+	return &tomlDecoder{
+		finished: false,
+		d:        NewDataTreeNavigator(),
+	}
+}
+
+func (dec *tomlDecoder) Init(reader io.Reader) error {
+	dec.parser = toml.Parser{}
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(reader)
+	dec.parser.Reset(buf.Bytes())
+	return nil
+}
+
+func (dec *tomlDecoder) getFullPath(tomlNode *toml.Node) ([]interface{}, error) {
+	path := make([]interface{}, 0)
+	for {
+		path = append(path, string(tomlNode.Data))
+		tomlNode = tomlNode.Next()
+		if tomlNode == nil {
+			return path, nil
+		}
+	}
+}
+
+func (dec *tomlDecoder) createKeyValueMap(tomlNode *toml.Node) (*yaml.Node, error) {
+
+	value := tomlNode.Value()
+	path, err := dec.getFullPath(value.Next())
+	if err != nil {
+		return nil, err
+	}
+
+	rootMap := &CandidateNode{
+		Node: &yaml.Node{
+			Kind: yaml.MappingNode,
+			Tag:  "!!map",
+		},
+	}
+
+	valueNode, err := dec.convertToYamlNode(value)
+	if err != nil {
+		return nil, err
+	}
+	context := Context{}
+	context = context.SingleChildContext(rootMap)
+
+	err = dec.d.DeeplyAssign(context, path, valueNode)
+	if err != nil {
+		return nil, err
+	}
+
+	return rootMap.Node, nil
+}
+
+func (dec *tomlDecoder) createTable(tomlNode *toml.Node) (*yaml.Node, error) {
+	log.Debug("Table: %v", string(tomlNode.Data))
```

Comment:
here specifically - what's the toml.Node structure for the table? If you look at the test I've got in toml_test.go - I have
```
[owners]
blah = "X"
```

But I only see to get owners in the node child here 🤔

Reply:
Correct. The parser returns top level expressions. So you should have a first node for the table header, then another one for key-value.

Reply:
So in this example you need two calls to NextExpression() to parse the whole document (arguably 3 to detect that you're at the end of the document and stop).